File: HW/HW7/minMeetingRooms.py
from typing import List
import heapq

class ServerManager:

    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0
          
        intervals.sort(key=lambda x: x[0])

        end_times = []

        heapq.heappush(end_times, intervals[0][1])

        for i in range(1, len(intervals)):
            start_time = intervals[i][0]

            if end_times and start_time >= end_times[0]:
      
                heapq.heappop(end_times)
            heapq.heappush(end_times, intervals[i][1])

        return len(end_times)

# Test cases
server_manager = ServerManager()

intervals1 = [[0, 30], [5, 10], [15, 20]]
print(server_manager.minMeetingRooms(intervals1))  # Output: 2

intervals2 = [[0, 1], [1, 2], [2, 3]]
print(server_manager.minMeetingRooms(intervals2))

File: HW/HW7/minMeetingRooms.java
import java.util.Arrays;
import java.util.PriorityQueue;

public class ServerManager {

    public int minMeetingRooms(int[][] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }
          Arrays.sort(intervals, (a, b) -> a[0] - b[0]);

        PriorityQueue<Integer> endTimes = new PriorityQueue<>();
        endTimes.offer(intervals[0][1]);

        for (int i = 1; i < intervals.length; i++) {
            int startTime = intervals[i][0];

            if (!endTimes.isEmpty() && startTime >= endTimes.peek()) {
                endTimes.poll();
            }
            endTimes.offer(intervals[i][1]);
        }
        return endTimes.size();
    }

    public static void main(String[] args) {
        ServerManager serverManager = new ServerManager();

        int[][] intervals1 = {{0, 30}, {5, 10}, {15, 20}};
        System.out.println(serverManager.minMeetingRooms(intervals1));  

        int[][] intervals2 = {{0, 1}, {1, 2}, {2, 3}};
        System.out.println(serverManager.minMeetingRooms(intervals2)); 
    }
}
